### PR TITLE
refactor(select): merged ps-select-data into ps-select

### PR DESCRIPTION
**Pull request by saithis**
This also fixes the state of the toggle all checkbox and possibly a rare condition where mat-select tries to access the options before they are set.

### CODE REVIEW

File: projects/components/select/src/select.component.html [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        [formControl]="filterCtrl"
+        [searching]="loading"
+        [showToggleAllCheckbox]="showToggleAll"
+        [toggleAllCheckboxChecked]="toggleAllCheckboxChecked"
+        [toggleAllCheckboxIndeterminate]="toggleAllCheckboxIndeterminate"
+        (toggleAll)="onToggleAll($event)"
+      ></ngx-mat-select-search>
+    </mat-option>
+
+    <mat-option *ngIf="hasError" [disabled]="true" class="ps-select__error">
```
Comment (6rayWolf) [inline projects/components/select/src/select.component.html:31]: There is no definition for this class in the scss

File: projects/components/select/src/testing/select.harness.ts [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    return null;
+  }
+
+  /** Gets the option that shows the error message. */
+  async getErrorOption() {
+    const options = await this.getOptions();
+    for (const option of options) {
+      if (await (await option.host()).hasClass('ps-select__error')) {
```
Comment (6rayWolf) [inline projects/components/select/src/testing/select.harness.ts:152]: I couldn't find any definition for this class? Is this just a marker class?

File: projects/components/select/src/select.component.ts [hunk 3]
```diff
@@ -1,53 +1,70 @@
...
+import { BehaviorSubject, Subject, Subscription } from 'rxjs';
+import { takeUntil } from 'rxjs/operators';
+import { getSelectUnknownDataSourceError } from './errors';
+import { PsSelectItem } from './models';
+import { DEFAULT_COMPARER, isPsSelectDataSource, PsSelectDataSource } from './select-data-source';
 import { PsSelectOptionTemplateDirective } from './select-option-template.directive';
 import { PsSelectTriggerTemplateDirective } from './select-trigger-template.directive';
+import { PsSelectService } from './select.service';
+
+const enum ValueChangeSource {
```
Comment (6rayWolf) [inline projects/components/select/src/select.component.ts:40]: unused?

**Review by 6rayWolf — CHANGES_REQUESTED** [review]

File: projects/components/select/src/select.component.ts (hunk 3 above)
Comment (saithis) [reply projects/components/select/src/select.component.ts:40]: it is used in _propagateValueChange

File: projects/components/select/src/testing/select.harness.ts (hunk 2 above)
Comment (saithis) [reply projects/components/select/src/testing/select.harness.ts:152]: yes, same for ps-select__search

File: projects/components/select/src/select.component.html (hunk 1 above)
Comment (saithis) [reply projects/components/select/src/select.component.html:31]: it's a marker class

File: projects/components/select/src/select.component.html [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    <mat-option *ngIf="showEmptyInput" class="ps-select__empty-option">
+      --
+    </mat-option>
+
+    <mat-option
+      *ngFor="let item of items; trackBy: trackByOptions"
+      [value]="item.value"
+      [class.ps-option-hidden]="item.hidden"
+      [disabled]="item.disabled"
+      class="ps-select__option"
```
[the remark this answers is not in the document]
Comment (saithis) [reply projects/components/select/src/select.component.html:44]: it's a marker class to differentiate these options from the others (filter, error, ...)

File: projects/components/select/src/select.component.ts [hunk 5]
```diff
@@ -229,40 +267,127 @@ export class PsSelectComponent<T = any> implements ControlValueAccessor, MatForm
...
+    return `${item.value}#${item.label}`;
+  }
+
+  private _propagateValueChange(value: any, source: ValueChangeSource) {
+    this._value = value;
+    this.empty = this.multiple ? !value?.length : value == null;
+    this._updateToggleAllCheckbox();
+    this._pushSelectedValuesToDataSource(value);
+    if (source !== ValueChangeSource.ValueInput) {
+      this.valueChange.emit(value); // nicht wenn über input
```
[the remark this answers is not in the document]
Comment (saithis) [reply projects/components/select/src/select.component.ts:343]: yes

File: projects/components/select/src/select.component.ts [hunk 6]
```diff
@@ -229,40 +267,127 @@ export class PsSelectComponent<T = any> implements ControlValueAccessor, MatForm
...
+    }
+    if (source !== ValueChangeSource.WriteValue) {
+      this._onChange(value); // nicht wenn über writevalue
```
[the remark this answers is not in the document]
Comment (saithis) [reply projects/components/select/src/select.component.ts:346]: dito

File: projects/components/select/src/select.component.scss [hunk 7]
```diff
@@ -0,0 +1,6 @@
+.ps-option-hidden {
```
[the remark this answers is not in the document]
Comment (saithis) [reply projects/components/select/src/select.component.scss:1]: i'll change it to ps-select__option--hidden

File: projects/components/select/src/select.component.html [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    [required]="required"
+    (valueChange)="onValueChange($event)"
+    (selectionChange)="onSelectionChange($event)"
+    (openedChange)="onOpenedChange($event)"
+  >
+    <mat-select-trigger *ngIf="triggerTemplate && !empty">
+      <ng-container [ngTemplateOutlet]="triggerTemplate" [ngTemplateOutletContext]="{ $implicit: customTriggerData }"></ng-container>
+    </mat-select-trigger>
+
+    <mat-option class="ps-select__search">
```
Comment (Yirica) [inline projects/components/select/src/select.component.html:20]: No definition for ps-select_search either

File: projects/components/select/src/select.component.html [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        [toggleAllCheckboxIndeterminate]="toggleAllCheckboxIndeterminate"
+        (toggleAll)="onToggleAll($event)"
+      ></ngx-mat-select-search>
+    </mat-option>
+
+    <mat-option *ngIf="hasError" [disabled]="true" class="ps-select__error">
+      <span class="ps-select__error-message">{{ error | psErrorMessage }}</span>
+    </mat-option>
+
+    <mat-option *ngIf="showEmptyInput" class="ps-select__empty-option">
```
Comment (Yirica) [inline projects/components/select/src/select.component.html:35]: No definition for ps-select__empty-option either

File: projects/components/select/src/select.component.ts [hunk 10]
```diff
@@ -71,116 +89,118 @@ export class PsSelectComponent<T = any> implements ControlValueAccessor, MatForm
...
+  public get loading(): boolean {
+    return !!this._dataSourceInstance?.loading;
   }
-  public get formControl(): FormControl {
-    // event binding only -> dummy control
-    return (this.ngControl && (this.ngControl.control as FormControl)) || this._dummyControl;
+
+  /** true when there was an error while loading the options */
+  public get hasError(): boolean {
+    return !!this._dataSourceInstance?.error;
```
Comment (Yirica) [inline projects/components/select/src/select.component.ts:184]: Could be replaced by !!this.error;

File: projects/components/select/src/select.component.html (hunk 9 above)
Comment (saithis) [reply projects/components/select/src/select.component.html:35]: it's a marker class

File: projects/components/select/src/select.component.ts (hunk 10 above)
Comment (saithis) [reply projects/components/select/src/select.component.ts]: done

**Review by 6rayWolf — APPROVED** [review]

**Review by Yirica — APPROVED** [review]

**Pull request merged**